frontend/src/entities/dataset/dataset-drag-and-drop.tsx

```diff
@@ -0,0 +1,152 @@
+    }
+
+    function dragWindowEnterHandler (event) {
+        if (event.dataTransfer.types && (event.dataTransfer.types.indexOf ? event.dataTransfer.types.indexOf('Files') !== -1 : event.dataTransfer.types.contains('Files'))) {
+            document.getElementsByClassName('dropzone')[0].classList.remove('display-none');
+        }
+    }
+
+    useEffect(() => {
+        document.addEventListener('dragenter', dragWindowEnterHandler);
+
+        return () => {
+            document.removeEventListener('dragenter', dragWindowEnterHandler);
+        };
+    }, []);
+    
+    /**
+     * Asynchronously retrieves the files from the provided DataTransfer object.
+     * It handles both individual files and directories, reading the
+     * content of directories and returning all the files in the directory hierarchy.
+     *
+     * @param dataTransfer - The DataTransfer object containing the dragged files.
+     * @returns A Promise that resolves to an array of File objects.
+     */
+    async function getFilesAsync (dataTransfer: DataTransfer) {
+        const files: File[] = [];
+        for (let i = 0; i < dataTransfer.items.length; i++) {
+            const item = dataTransfer.items[i];
+            if (item.kind === 'file') {
+                if (typeof item.webkitGetAsEntry === 'function') {
+                    const entry = item.webkitGetAsEntry();
+                    const entryContent = await readEntryContentAsync(entry);
+                    files.push(...entryContent);
+                    continue;
+                }
+    
+                const file = item.getAsFile();
+                if (file) {
+                    files.push(file);
+                }
+            }
+        }
+        return files;
+    }
+
+    // Returns a promise with all the files of the directory hierarchy
+    function readEntryContentAsync (entry: FileSystemEntry) {
+        return new Promise<File[]>((resolve) => {
+            let reading = 0;
+            const contents: File[] = [];
+
+            readEntry(entry);
+
+            function readEntry (entry: FileSystemEntry) {
+                if (isFile(entry)) {
+                    reading++;
+                    entry.file((file) => {
+                        reading--;
+                        contents.push(file);
+
+                        if (reading === 0) {
+                            resolve(contents);
+                        }
+                    });
+                } else if (isDirectory(entry)) {
+                    readReaderContent(entry.createReader());
+                }
+            }
+
+            function readReaderContent (reader: FileSystemDirectoryReader) {
+                reading++;
+
+                reader.readEntries(function (entries) {
+                    reading--;
+                    for (const entry of entries) {
+                        readEntry(entry);
+                    }
+
+                    if (reading === 0) {
```

Comment:
What happens if `reading` != 0 by the end? Also what's the `resolve` keyword doing a few lines below, just trying to understand what this `reading` variable is doing and its logic.

Reply:
When reading !== 0, it means that there are still some files being read asynchronously. This can happen when the function is reading the contents of a directory hierarchy, where it encounters subdirectories and needs to read the files within those subdirectories as well.